cpp/velox/substrait/SubstraitToVeloxPlan.cc

```diff
@@ -793,23 +794,34 @@ const core::WindowNode::Frame createWindowFrame(
       VELOX_FAIL("the window type only support ROWS and RANGE, and the input type is ", std::to_string(type));
   }
 
-  auto boundTypeConversion = [](::substrait::Expression_WindowFunction_Bound boundType)
+  auto boundTypeConversion = [&](::substrait::Expression_WindowFunction_Bound boundType)
       -> std::tuple<core::WindowNode::BoundType, core::TypedExprPtr> {
-    // TODO: support non-literal expression.
     if (boundType.has_current_row()) {
       return std::make_tuple(core::WindowNode::BoundType::kCurrentRow, nullptr);
     } else if (boundType.has_unbounded_following()) {
       return std::make_tuple(core::WindowNode::BoundType::kUnboundedFollowing, nullptr);
     } else if (boundType.has_unbounded_preceding()) {
       return std::make_tuple(core::WindowNode::BoundType::kUnboundedPreceding, nullptr);
     } else if (boundType.has_following()) {
-      return std::make_tuple(
-          core::WindowNode::BoundType::kFollowing,
-          std::make_shared<core::ConstantTypedExpr>(BIGINT(), variant(boundType.following().offset())));
+      if (boundType.following().has_offset()) {
+        return std::make_tuple(
+            core::WindowNode::BoundType::kFollowing,
+            std::make_shared<core::ConstantTypedExpr>(BIGINT(), variant(boundType.following().offset())));
+      } else {
+        return std::make_tuple(
+            core::WindowNode::BoundType::kFollowing,
+            exprConverter_->toVeloxExpr(boundType.following().ref(), inputType));
+      }
     } else if (boundType.has_preceding()) {
-      return std::make_tuple(
-          core::WindowNode::BoundType::kPreceding,
-          std::make_shared<core::ConstantTypedExpr>(BIGINT(), variant(boundType.preceding().offset())));
+      if (boundType.preceding().has_offset()) {
+        return std::make_tuple(
+            core::WindowNode::BoundType::kPreceding,
+            std::make_shared<core::ConstantTypedExpr>(BIGINT(), variant(boundType.preceding().offset())));
+      } else {
+        return std::make_tuple(
+            core::WindowNode::BoundType::kPreceding,
+            exprConverter_->toVeloxExpr(boundType.preceding().ref(), inputType));
+      }
```

Comment:
ditto

Reply:
done. 
I tried to reuse the logic here, but since the type of `boundType.preceding()` and `boundType.following()` are different, so I have to pass the `offset` and `ref` instead of the expression itself